src/v/cluster/controller_backend.cc

```diff
@@ -473,6 +476,43 @@ ss::future<> controller_backend::fetch_deltas() {
       });
 }
 
+ss::future<> controller_backend::clear_orphan_topic_files() {
+    if (ss::this_shard_id() != 0) {
+        return ss::now();
+    }
+    auto topics = _topics.local().all_topics();
+    storage::log_manager::ntp_to_revision topics_on_node;
+    // Init with default namespace to clean if there is no topics
+    absl::flat_hash_set<model::ns> namespaces = {{model::ns("kafka")}};
+    for (const auto& t : topics) {
+        namespaces.emplace(t.ns);
+        auto meta = _topics.local().get_topic_assignments(t);
+        if (meta) {
+            for (const auto& p : *meta) {
+                model::ntp ntp(t.ns, t.tp, p.id);
+                auto revision = _topics.local()
+                                  .get_topic_metadata(
+                                    model::topic_namespace_view(ntp))
+                                  ->get_revision();
+                auto previous_replica_set
+                  = _topics.local().get_previous_replica_set(ntp);
+                if (
+                  contains_node(_self, p.replicas)
+                  || (previous_replica_set && contains_node(_self, *previous_replica_set))) {
+                    topics_on_node.emplace(ntp, revision);
+                }
+            }
+        }
+    }
+    auto last_applied_revision = _topics.local().last_applied_revision();
```

Comment:
last_applied_revision is updated as topic table applies the deltas. So by the time this cleanup runs, it may not have applied all the deltas, so we potentially may be operating on a not-yet-up-to-date snapshot of the topics, no?

Reply:
After controller backend bootstrap we must have reconcilied all messages in `controller_log` and `topic_table` deltas that were on node before restart.
Orphan files might be left only for topics that existed before we have restarted node. So all deltas for topics that may have orphan files must be reconcilied at this point.
If we running deletion of topic files while node is running we will retry it until success. So we cannot get orphan files in runtime.